In [1]:
import plotly_express as px
import pandas as pd

# Importing dataset from excel file

In [14]:
df_vaccine_data = pd.read_excel("./Data/folkhalsomyndigheten_Covid19_Vaccine.xlsx", sheet_name= "Vaccinerade kommun och ålder")

In [15]:
df_vaccine_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2900 entries, 0 to 2899
Data columns (total 14 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Län                  2900 non-null   int64  
 1   Län_namn             2900 non-null   object 
 2   Kommun               2900 non-null   int64  
 3   Kommun_namn          2900 non-null   object 
 4   Ålder                2900 non-null   object 
 5   Befolkning           2900 non-null   int64  
 6   Antal minst 1 dos    2900 non-null   int64  
 7   Antal minst 2 doser  2900 non-null   int64  
 8   Antal 3 doser        2320 non-null   float64
 9   Antal 4 doser        870 non-null    float64
 10  Andel minst 1 dos    2900 non-null   float64
 11  Andel minst 2 doser  2900 non-null   float64
 12  Andel 3 doser        2320 non-null   float64
 13  Andel 4 doser        870 non-null    float64
dtypes: float64(6), int64(5), object(3)
memory usage: 317.3+ KB


In [16]:
df_vaccine_data.head(7)

Län        Län_namn  Kommun     Kommun_namn  Ålder  Befolkning  \
0    1  Stockholms län     114  Upplands Väsby  12-15        2422   
1    1  Stockholms län     114  Upplands Väsby  16-17        1203   
2    1  Stockholms län     114  Upplands Väsby  18-29        6692   
3    1  Stockholms län     114  Upplands Väsby  30-39        7332   
4    1  Stockholms län     114  Upplands Väsby  40-49        6946   
5    1  Stockholms län     114  Upplands Väsby  50-59        6332   
6    1  Stockholms län     114  Upplands Väsby  60-69        4258   

   Antal minst 1 dos  Antal minst 2 doser  Antal 3 doser  Antal 4 doser  \
0               1206                 1046            NaN            NaN   
1                839                  755            NaN            NaN   
2               4887                 4469         1959.0            NaN   
3               5542                 5240         2878.0            NaN   
4               5592                 5429         3719.0            NaN   
5               5644                 5558         4535.0            NaN   
6               3829                 3786         3335.0            NaN   

   Andel minst 1 dos  Andel minst 2 doser  Andel 3 doser  Andel 4 doser  
0           0.497936             0.431874            NaN            NaN  
1           0.697423             0.627598            NaN            NaN  
2           0.730275             0.667812       0.292738            NaN  
3           0.755865             0.714675       0.392526            NaN  
4           0.805068             0.781601       0.535416            NaN  
5           0.891346             0.877764       0.716203            NaN  
6           0.899248             0.889150       0.783232            NaN

### Task A
Counties in the dataset

In [17]:
df_vaccine_data["Län_namn"].unique()

array(['Stockholms län', 'Uppsala län', 'Södermanlands län',
       'Östergötlands län', 'Jönköpings län', 'Kronobergs län',
       'Kalmar län', 'Gotlands län', 'Blekinge län', 'Skåne län',
       'Hallands län', 'Västra Götalands län', 'Värmlands län',
       'Örebro län', 'Västmanlands län', 'Dalarnas län', 'Gävleborgs län',
       'Västernorrlands län', 'Jämtlands län', 'Västerbottens län',
       'Norrbottens län'], dtype=object)

In [18]:
df_vaccine_data["Län_namn"].nunique()

21

There are 21 counties in sweden, so the dataset seems to have data about all of them.

### Task B
Municipalities in the dataset

In [22]:
df_vaccine_data["Kommun_namn"].unique()

array(['Upplands Väsby', 'Vallentuna', 'Österåker', 'Värmdö', 'Järfälla',
       'Ekerö', 'Huddinge', 'Botkyrka', 'Salem', 'Haninge', 'Tyresö',
       'Upplands-Bro', 'Nykvarn', 'Täby', 'Danderyd', 'Sollentuna',
       'Stockholm', 'Södertälje', 'Nacka', 'Sundbyberg', 'Solna',
       'Lidingö', 'Vaxholm', 'Norrtälje', 'Sigtuna', 'Nynäshamn', 'Håbo',
       'Älvkarleby', 'Knivsta', 'Heby', 'Tierp', 'Uppsala', 'Enköping',
       'Östhammar', 'Vingåker', 'Gnesta', 'Nyköping', 'Oxelösund', 'Flen',
       'Katrineholm', 'Eskilstuna', 'Strängnäs', 'Trosa', 'Ödeshög',
       'Ydre', 'Kinda', 'Boxholm', 'Åtvidaberg', 'Finspång',
       'Valdemarsvik', 'Linköping', 'Norrköping', 'Söderköping', 'Motala',
       'Vadstena', 'Mjölby', 'Aneby', 'Gnosjö', 'Mullsjö', 'Habo',
       'Gislaved', 'Vaggeryd', 'Jönköping', 'Nässjö', 'Värnamo', 'Sävsjö',
       'Vetlanda', 'Eksjö', 'Tranås', 'Uppvidinge', 'Lessebo', 'Tingsryd',
       'Alvesta', 'Älmhult', 'Markaryd', 'Växjö', 'Ljungby', 'Högsby',
       '

In [23]:
df_vaccine_data["Kommun_namn"].nunique()

290

There are 290 counties in sweden, so the dataset seems to have data about all of them as well.

### Task C
Population in the dataset

In [33]:
vaccine_data_population = df_vaccine_data["Befolkning"].sum()

In [34]:
vaccine_data_population

9092790

It seems that the population in the dataset is 9 092 790 people.

### Task D
Children under age 18

In [30]:
df_vaccine_data_ages = df_vaccine_data["Ålder"].unique()

In [31]:
df_vaccine_data_ages

array(['12-15', '16-17', '18-29', '30-39', '40-49', '50-59', '60-69',
       '70-79', '80-89', '90 eller äldre'], dtype=object)

Looking at the dataset, it doesn't seem to have data about children under the age of 12.
We'll have to use info about whole population of sweden in 2022 to help us find those.
According to SCB the population of sweden was 10 521 556 in 2022.

In [32]:
sweden_population_2022 = 10521556

In [36]:
excluded_age_groups = ["12-15", "16-17"]
adult_population = df_vaccine_data[~df_vaccine_data["Ålder"].isin(excluded_age_groups)]["Befolkning"].sum()

In [37]:
adult_population

8347420

In [38]:
child_population = sweden_population_2022 - adult_population

In [39]:
child_population

2174136

Using the data from SCB we can find out that the population of children under the age of 18 is 2 174 136.

### Task E
Age distribution diagram.

We'll need to add the population of ages 0-11 to the dataset before we can make the diagram.

In [118]:
ages_0_11 = {"Ålder" : "0-11", "Befolkning" : sweden_population_2022 - vaccine_data_population}

In [119]:
df_ages_0_11 = pd.DataFrame([ages_0_11])

In [120]:
df_ages_0_11

Ålder  Befolkning
0  0-11     1428766

In [121]:
df_age_ranges_population = pd.DataFrame({"Ålder" : df_vaccine_data_ages, "Befolkning" : 0})

In [122]:
df_age_ranges_population

Ålder  Befolkning
0           12-15           0
1           16-17           0
2           18-29           0
3           30-39           0
4           40-49           0
5           50-59           0
6           60-69           0
7           70-79           0
8           80-89           0
9  90 eller äldre           0

In [123]:
df_age_ranges_population["Befolkning"] = df_vaccine_data.groupby("Ålder")["Befolkning"].sum().values

In [124]:
df_age_ranges_population

Ålder  Befolkning
0           12-15      503831
1           16-17      241539
2           18-29     1475950
3           30-39     1467590
4           40-49     1298156
5           50-59     1339798
6           60-69     1121922
7           70-79     1033113
8           80-89      496750
9  90 eller äldre      114141

In [127]:
df_age_ranges_population = pd.concat([df_ages_0_11, df_age_ranges_population], ignore_index=True)

In [128]:
df_age_ranges_population

Ålder  Befolkning
0             0-11     1428766
1            12-15      503831
2            16-17      241539
3            18-29     1475950
4            30-39     1467590
5            40-49     1298156
6            50-59     1339798
7            60-69     1121922
8            70-79     1033113
9            80-89      496750
10  90 eller äldre      114141

Creating a pie-chart to show the population by age group.

In [134]:
category_order = df_age_ranges_population["Ålder"].tolist()

age_group_population_chart = px.pie(
    df_age_ranges_population, names="Ålder", values="Befolkning", 
    title="Swedish population by age groups", category_orders={"Ålder": category_order}
)
age_group_population_chart.show()
age_group_population_chart.write_html("./Visualisations/Task_2_plotly_swedish_population_by_age_group.html")

### Task F
Diagram with amount of doses

In [169]:
df_vaccine_data.head(10)

Län        Län_namn  Kommun     Kommun_namn           Ålder  Befolkning  \
0    1  Stockholms län     114  Upplands Väsby           12-15        2422   
1    1  Stockholms län     114  Upplands Väsby           16-17        1203   
2    1  Stockholms län     114  Upplands Väsby           18-29        6692   
3    1  Stockholms län     114  Upplands Väsby           30-39        7332   
4    1  Stockholms län     114  Upplands Väsby           40-49        6946   
5    1  Stockholms län     114  Upplands Väsby           50-59        6332   
6    1  Stockholms län     114  Upplands Väsby           60-69        4258   
7    1  Stockholms län     114  Upplands Väsby           70-79        3987   
8    1  Stockholms län     114  Upplands Väsby           80-89        1934   
9    1  Stockholms län     114  Upplands Väsby  90 eller äldre         337   

   Antal minst 1 dos  Antal minst 2 doser  Antal 3 doser  Antal 4 doser  \
0               1206                 1046            NaN            NaN   
1                839                  755            NaN            NaN   
2               4887                 4469         1959.0            NaN   
3               5542                 5240         2878.0            NaN   
4               5592                 5429         3719.0            NaN   
5               5644                 5558         4535.0            NaN   
6               3829                 3786         3335.0            NaN   
7               3786                 3758         3584.0         3051.0   
8               1859                 1856         1802.0         1617.0   
9                330                  327          319.0          277.0   

   Andel minst 1 dos  Andel minst 2 doser  Andel 3 doser  Andel 4 doser  
0           0.497936             0.431874       0.000000       0.000000  
1           0.697423             0.627598       0.000000       0.000000  
2           0.730275             0.667812       0.292738       0.000000  
3           0.755865             0.714675       0.392526       0.000000  
4           0.805068             0.781601       0.535416       0.000000  
5           0.891346             0.877764       0.716203       0.000000  
6           0.899248             0.889150       0.783232       0.000000  
7           0.949586             0.942563       0.898921       0.765237  
8           0.961220             0.959669       0.931748       0.836091  
9           0.979228             0.970326       0.946588       0.821958

In [170]:
df_vaccine_data.isna().sum()

Län                       0
Län_namn                  0
Kommun                    0
Kommun_namn               0
Ålder                     0
Befolkning                0
Antal minst 1 dos         0
Antal minst 2 doser       0
Antal 3 doser           580
Antal 4 doser          2030
Andel minst 1 dos         0
Andel minst 2 doser       0
Andel 3 doser             0
Andel 4 doser             0
dtype: int64

There are quite a few NaN values in the dataset that we need to clean.
(also cleaning "Andel 4 doser" since it will be used in the next task)

In [192]:
df_vaccine_data["Andel 3 doser"] = df_vaccine_data["Andel 3 doser"].fillna(0)
df_vaccine_data["Andel 4 doser"] = df_vaccine_data["Andel 4 doser"].fillna(0)
df_vaccine_data.isna().sum()

Län                       0
Län_namn                  0
Kommun                    0
Kommun_namn               0
Ålder                     0
Befolkning                0
Antal minst 1 dos         0
Antal minst 2 doser       0
Antal 3 doser           580
Antal 4 doser          2030
Andel minst 1 dos         0
Andel minst 2 doser       0
Andel 3 doser             0
Andel 4 doser             0
dtype: int64

In [193]:
grouped_counties = df_vaccine_data.groupby("Län_namn")
vaccinated_per_county = grouped_counties[
                            ["Andel minst 1 dos", "Andel minst 2 doser", "Andel 3 doser", "Andel 4 doser"]
                        ].mean() *100

In [194]:
vaccinated_per_county = vaccinated_per_county.reset_index()
vaccinated_per_county

Län_namn  Andel minst 1 dos  Andel minst 2 doser  \
0           Blekinge län          85.645562            84.199263   
1           Dalarnas län          85.555434            84.134352   
2           Gotlands län          89.034008            87.341121   
3         Gävleborgs län          85.168444            82.888678   
4           Hallands län          85.923815            84.421520   
5          Jämtlands län          86.092991            84.396578   
6         Jönköpings län          86.004452            84.670980   
7             Kalmar län          85.129261            83.946415   
8         Kronobergs län          82.498228            81.019738   
9        Norrbottens län          87.033895            85.537685   
10             Skåne län          83.586733            81.909372   
11        Stockholms län          83.586828            80.593619   
12     Södermanlands län          85.791380            83.344720   
13           Uppsala län          88.242125            86.462405   
14         Värmlands län          88.083939            86.878431   
15     Västerbottens län          88.777276            87.432633   
16   Västernorrlands län          87.226690            85.030002   
17      Västmanlands län          83.648946            81.411815   
18  Västra Götalands län          85.942073            83.968784   
19            Örebro län          85.721485            84.344199   
20     Östergötlands län          87.930850            86.861533   

    Andel 3 doser  Andel 4 doser  
0       59.592201      24.815388  
1       59.475201      26.436745  
2       62.246828      26.338467  
3       57.957835      25.300869  
4       59.935017      25.794072  
5       57.845140      24.872016  
6       59.975818      25.922680  
7       60.302130      25.815887  
8       55.650754      24.569816  
9       60.227573      25.444300  
10      55.866455      24.636018  
11      56.832073      24.225738  
12      58.104702      25.588125  
13      61.398145      26.374325  
14      61.601844      26.734018  
15      61.853894      27.123339  
16      60.007273      25.014668  
17      57.739257      25.648359  
18      57.790836      25.341537  
19      59.917004      25.457560  
20      62.376387      26.562731

In [203]:
amount_of_doses_county_diagram = px.bar(vaccinated_per_county,
                                x="Län_namn", 
                                y=["Andel minst 1 dos", "Andel minst 2 doser", "Andel 3 doser"],
                                title="Percentage amount of doses given by County",
                                labels={"value": "Percentage", "variable": "Doses (%)", "Län_namn": "Län"},
                                barmode="group")
amount_of_doses_county_diagram.show()
amount_of_doses_county_diagram.write_html("./Visualisations/Task_2_plotly_amount_of_doses_per_county.html")

### Task G
Diagram with amount of doses for specific counties

In [204]:
included_counties = ["Stockholms län", "Västra Götalands län"]
stholm_and_vgotaland_counties = vaccinated_per_county[vaccinated_per_county["Län_namn"].isin(included_counties)]

In [205]:
stholm_and_vgotaland_counties

Län_namn  Andel minst 1 dos  Andel minst 2 doser  \
11        Stockholms län          83.586828            80.593619   
18  Västra Götalands län          85.942073            83.968784   

    Andel 3 doser  Andel 4 doser  
11      56.832073      24.225738  
18      57.790836      25.341537

In [207]:
stholm_and_vgotaland_counties_diagram = px.bar(stholm_and_vgotaland_counties, 
                                               x="Län_namn", 
                                               y=["Andel minst 1 dos", 
                                                  "Andel minst 2 doser", 
                                                  "Andel 3 doser",
                                                  "Andel 4 doser"],
                                               title="Percentage amount of doses given, Stockholm and Västra Götaland",
                                               labels={"value": "Percentage", "variable": "Doses (%)", "Län_namn": "Län"},
                                               barmode="group"
                                               )
stholm_and_vgotaland_counties_diagram.show()
stholm_and_vgotaland_counties_diagram.write_html("./Visualisations/Task_2_plotly_amount_of_doses_stholm_and_vgotaland.html")
